### ISSUE TITLE
BUG: Concurrency in prepareForUse() in container signature ByteSequences.

### ISSUE THREAD

**Issue by nishihatapalmer** [digital-preservation/droid#307]
ByteSequences need to be prepared for use before they can be used.

Binary signatures are prepared for use after the signature file is read, and before they are made available to different threads.

However, container signatures are compiled on-demand when they are first encountered.  This can happen from multiple threads, and it leads to multiple compilations of the same container signature.

**Comment by nishihatapalmer** [comment]
I'm not an expert in concurrency, but I've created the following fix which seems to work - but I'm not confident it's correct.

1.  There is a volatile "preparedForUse" boolean flag.
2.  This gets set to true when it's prepared.
3.  Preparation only happens when the flag is false, and is set to true after it happens the first time.

The effect seems to be that the ByteSequence is only prepared for use once.  However, I'm dubious this is enough for true protection.  If two executions happen simultaneously, they can both process the ByteSequence before the flag is set to true.

**Comment by nishihatapalmer** [comment]
I'm also not confident that the results of preparing for use in one thread will be available to the others.

**Comment by nishihatapalmer** [comment]
There is an argument that we should just compile all the container signatures up front before we pass them to multiple threads.  It saves a tiny amount of time to not compile a few hundred signatures (milliseconds), vs. the cost of maintaining synchronisation over on-demand compilation in multiple threads...

**Comment by nishihatapalmer** [comment]
I've tried adding synchronized to getCompiledBinarySignatures() in ContainerFile.

This seems to stop multiple compilations of the same container signature - but I can't swear it's safe.  Can't see any reason why there might be a deadlock or too much contention.

But...  all of this build-signatures-on-demand is supposed to make things a bit snappier.  

It would be much snappier overall if we just compiled them up front when the signature file loads (just like binary signatures).  Then we don't need any synchronisation or have threading issues.
